tools/SourceKit/tools/sourcekitd/lib/API/sourcekitdAPI-InProc.cpp

```diff
@@ -0,0 +1,969 @@
+//===--- sourcekitdAPI-InProc.cpp -----------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2016 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+#include "DictionaryKeys.h"
+#include "sourcekitd/sourcekitd.h"
+#include "sourcekitd/Internal.h"
+#include "sourcekitd/CodeCompletionResultsArray.h"
+#include "sourcekitd/DocSupportAnnotationArray.h"
+#include "sourcekitd/TokenAnnotationsArray.h"
+#include "sourcekitd/Logging.h"
+#include "SourceKit/Support/UIdent.h"
+#include "llvm/ADT/ArrayRef.h"
+#include "llvm/ADT/StringRef.h"
+#include "llvm/Support/ErrorHandling.h"
+#include "llvm/Support/MemoryBuffer.h"
+
+#include <map>
+
+using namespace SourceKit;
+using namespace sourcekitd;
+using llvm::ArrayRef;
+using llvm::StringRef;
+using llvm::raw_ostream;
+
+namespace {
+class SKDDictionary;
+class SKDArray;
+class SKDCustomData;
+class SKDError;
+
+class SKDObject {
+public:
+  SKDObject(): RetainCount(0) {}
+  virtual ~SKDObject() {}
+
+  SKDObject(SKDObject const&) = delete;
+  SKDObject &operator=(SKDObject const&) = delete;
+
+  SKDObject *retain() {
+    ++RetainCount;
+    return this;
+  }
+
+  void release() {
+    if (RetainCount == 0) {
+      delete this;
+    } else {
+      --RetainCount;
+    }
+  }
+
```

Comment:
llvm has helpers to define and use ref-counted objects, see "llvm/ADT/IntrusiveRefCntPtr.h" and "swift/Basic/ThreadSafeRefCounted.h" that introduces ThreadSafeRefCountedBaseVPTR.
Essentially you can do this:

`class SKDObject : public ThreadSafeRefCountedBaseVPTR {`

This will give you an object (with a virtual destructor) that can be passed around as ref-counted pointer using `llvm::IntrusiveRefCntPtr<T>`.